File: assignments/02-SDH/java/drone-remote-unit/src/main/java/it/unibo/sdh/impl/model/StateHolder.java
package it.unibo.sdh.impl.model;

import java.util.Optional;

public class StateHolder<S> {
    private Optional<S> state;
    
    public StateHolder() {
        setState(Optional.empty());
    }

    public StateHolder(S initialState) {
        setState(Optional.of(initialState));
    }

    public Optional<S> getCurrentState() {
        return state;
    }

    public void setState(final Optional<S> newState) {
        this.state = newState;
    }

    public void setState(final S newState) {
        setState(Optional.of(newState));
    }
}

File: assignments/02-SDH/arduino/smart-hangar/src/model/StateHolder.cpp
#include "StateHolder.hpp"

template <typename S>
StateHolder<S>::StateHolder(S initialState) {
    setState(initialState);
}

template <typename S>
S StateHolder<S>::getState() {
    return state;
}

template <typename S>
void StateHolder<S>::setState(S newState) {
    this->state = newState;
}